controllers/machine/machine_controller.go

```diff
@@ -0,0 +1,194 @@
+					continue
+				}
+				if newestEvent.LastTimestamp.Time.Before(event.LastTimestamp.Time) {
+					newestEvent = event
+					continue
+				}
+			}
+		}
+	}
+
+	return newestEvent
+}
+
+func parsePodsAndNamespacesFromEvent(reqLogger logr.Logger, event *corev1.Event) map[string]string {
+	re := regexp.MustCompile(`pods\/"([\w-]+)" -n "([\w-]+)"`)
+	matches := re.FindAllStringSubmatch(event.Message, -1)
+
+	// store the pod/namespace combinations with the pod as the key because
+	// the pod name _should_ generally be unique, where there may be multiple pods
+	// in each namespace
+	podNamespaces := map[string]string{}
+
+	for _, podMatch := range matches {
+		if len(podMatch) != 3 {
+			// I don't think this should ever happen, but this prevents trying to access indexes
+			// in the match slice that may not exist
+			reqLogger.Error(fmt.Errorf("Could not get the appropriate amount of matches"), "match", podMatch)
+			continue
+		}
+		// From the regex match we'll always get this podMatch slice with the following format:
+		// ['pods/"myPod-aaabbb" -n "namespace"', 'myPod-aaabbb', 'namespace']
+		podName := podMatch[1]
+		podNamespace := podMatch[2]
+
+		namespaceRe := regexp.MustCompile(`openshift-.*`)
+		// We don't generally care about openshift namespaces for customer metrics
+		if !namespaceRe.MatchString(podNamespace) {
+			podNamespaces[podName] = podNamespace
+		}
+	}
+	return podNamespaces
+}
+
+func (r *MachineReconciler) evaluateDeletingMachine(ctx context.Context, machine *machinev1beta1.Machine) (ctrl.Result, error) {
+	reqLogger := logf.FromContext(ctx).WithName(logName)
+
+	// Check Deleting Timestamp. If it's been less than 15m we don't care, requeue for 5m.
+	deletedTime := machine.GetDeletionTimestamp().Time
+	now := time.Now()
+	if !deletedTime.Before(now.Add(-15 * time.Minute)) {
+		reqLogger.Info("Machine was not deleted long enough ago. Requeueing after 5m.")
+		return utils.RequeueAfter(5 * time.Minute)
+	}
+
+	reqLogger.Info("Evaluating Deleting Machine")
+
+	machineEventList := &corev1.EventList{}
+	// we only want the events related to the machine that we're reconciling on
+	err := r.Client.List(ctx, machineEventList, &client.ListOptions{Namespace: "openshift-machine-api", FieldSelector: fields.SelectorFromSet(fields.Set{"involvedObject.name": machine.Name})})
+	if err != nil {
+		reqLogger.Error(err, "Unable to query events for machine")
+		return utils.RequeueWithError(err)
+	}
+
+	event := getMostRecentDrainFailedEvent(reqLogger, machineEventList)
+	if event == nil {
+		reqLogger.Info("No events returned for this machine")
+		// Try again - if there's no drain failures then we just keep requeueing until the machine is deleted and this is a noop
+		return utils.RequeueAfter(5 * time.Minute)
+	}
+	// TODO - do we need to ensure that the event happened within a specific timeframe? Like, within the last 15 minutes or something?
+
+	podNamespaces := parsePodsAndNamespacesFromEvent(reqLogger, event)
```

Comment:
Why do you pass in the `reqLogger` instead of having the function return an `error`? I think it would make testing easier if we could just check for an error returned instead of parsing the logger output in the unit tests.

Reply:
Good point. Updated for this and the other function.